File: language/compiler/src/com/kubadziworski/parsing/visitor/expression/ConditionalExpressionVisitor.java
package com.kubadziworski.parsing.visitor.expression;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.ConditionalExpressionContext;
import jiva.parser.JivaParser.ExpressionContext;
import com.kubadziworski.domain.node.expression.ConditionalExpression;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.domain.node.expression.Value;
import com.kubadziworski.domain.CompareSign;
import com.kubadziworski.domain.type.BultInType;
import org.antlr.v4.runtime.misc.NotNull;

public class ConditionalExpressionVisitor extends JivaBaseVisitor<ConditionalExpression> {
    private final ExpressionVisitor expressionVisitor;

    public ConditionalExpressionVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public ConditionalExpression visitConditionalExpression(@NotNull ConditionalExpressionContext ctx) {
        ExpressionContext leftExpressionCtx = ctx.expression(0);
        ExpressionContext rightExpressionCtx = ctx.expression(1);
        Expression leftExpression = leftExpressionCtx.accept(expressionVisitor);
        Expression rightExpression = rightExpressionCtx != null ? rightExpressionCtx.accept(expressionVisitor) : new Value(BultInType.INT, "0");
        CompareSign cmpSign = ctx.cmp != null ? CompareSign.fromString(ctx.cmp.getText()) : CompareSign.NOT_EQUAL;
        return new ConditionalExpression(leftExpression, rightExpression, cmpSign);
    }
}

File: language/compiler/src/parsing/visitor/expression/ConditionalExpressionVisitor.kt
package jiva.parsing.visitor.expression

import jiva.domain.CompareSign
import jiva.domain.node.expression.ConditionalExpression
import jiva.domain.node.expression.Value
import jiva.domain.type.BultInType
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.ConditionalExpressionContext

class ConditionalExpressionVisitor(private val visitor: ExpressionVisitor) : JivaBaseVisitor<ConditionalExpression>() {

    override fun visitConditionalExpression(ctx: ConditionalExpressionContext): ConditionalExpression {
        val leftContext = ctx.expression(0)
        val rightContext = ctx.expression(1)
        val leftExpression = leftContext.accept(visitor)
        val rightExpression = if (rightContext != null) rightContext.accept(visitor) else Value(BultInType.INT, "0")
        val cmpSign = if (ctx.cmp != null) CompareSign.fromString(ctx.cmp.text) else CompareSign.NOT_EQUAL
        return ConditionalExpression(leftExpression!!, rightExpression!!, cmpSign)
    }

}